### PR TITLE
Implement datamodule for ecoacoustics dataset

### PR DESCRIPTION
**Pull request by thomkeh**
supersedes #71

### CODE REVIEW

File: conduit/data/datamodules/audio/base.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+
+@attr.define(kw_only=True)
+class CdtAudioDataModule(CdtDataModule):
+
+    root: Union[str, Path] = attr.field(kw_only=False)
+
+    @property
+    @abstractmethod
+    def _base_augmentations(self) -> A.Compose:
```
Comment (thomkeh) [inline conduit/data/datamodules/audio/base.py:24]: these augmentation properties were removed a couple of weeks ago and replaced with default_transforms, with the transforms now able to be freely specified through the init.

Comment (thomkeh) [reply conduit/data/datamodules/audio/base.py:24]: That aside, the properties should return nn.Sequential objects, not A.Compose (which is image-specific).

Comment (MylesBartlett) [reply conduit/data/datamodules/audio/base.py:24]: Aren't these my words?

Comment (DavidHurst) [reply conduit/data/datamodules/audio/base.py:24]: I have heard this before, is this essential or can it be done another time?

Comment (MylesBartlett) [reply conduit/data/datamodules/audio/base.py:24]: You can just remove all the augmentation methods; they're not needed at the moment. _setup can also be removed as it isn't doing anything that the superclass's isn't.

File: conduit/data/datamodules/audio/ecoacoustics.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+__all__ = ["EcoacousticsDataModule"]
+
+
+@attr.define(kw_only=True)
+class EcoacousticsDataModule(CdtAudioDataModule):
+    """Data-module for the Ecoacoustics dataset."""
+
+    image_size: int = 224
+    class_train_props: Optional[dict] = None
```
Comment (MylesBartlett) [inline conduit/data/datamodules/audio/ecoacoustics.py:28]:
```suggestion
```

File: conduit/data/datamodules/audio/ecoacoustics.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        return A.Compose(
+            [
+                A.Resize(self.image_size, self.image_size),
+            ]
+        )
+
+    @property  # type: ignore[misc]
+    @implements(CdtAudioDataModule)
+    def _train_augmentations(self) -> A.Compose:
+        return A.Compose([])
```
Comment (MylesBartlett) [inline conduit/data/datamodules/audio/ecoacoustics.py:48]:
```suggestion
```

File: conduit/data/datamodules/audio/base.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        self._val_data = val
+        self._test_data = test
+
+    def _augmentations(self, train: bool) -> A.Compose:
+        # Base augmentations (augmentations that are applied to all splits of the data)
+        augs: list[AlbumentationsTform] = [self._base_augmentations]
+        # Add training augmentations on top of base augmentations
+        if train:
+            augs.append(self._train_augmentations)
+        return A.Compose(augs)
```
Comment (MylesBartlett) [inline conduit/data/datamodules/audio/base.py:45]:
```suggestion
```

File: conduit/data/datasets/audio/ecoacoustics.py [hunk 5]
```diff
@@ -98,8 +92,10 @@ def __init__(
         self.root = Path(root).expanduser()
         self.download = download
         self.base_dir = self.root / self._BASE_FOLDER
-        self.labels_dir = self.base_dir / self.INDICES_DIR
-        self._processed_audio_dir = self.base_dir / self._PROCESSED_DIR
+        self.labels_dir = self.base_dir / self._INDICES_FILE_INFO.name[:-4] / self.INDICES_DIR
+        # target directory needs to depend on the preprocessing function
+        preprocess_id = re.sub(r"\W+", '', str(preprocessing_transform))
```
Comment (MylesBartlett) [inline conduit/data/datasets/audio/ecoacoustics.py:97]:
```suggestion
        preprocess_id = preprocessing_transform.__class.__name__
```
Or something like this; perhaps you'd like to convert to snake case

File: conduit/data/datasets/audio/ecoacoustics.py [hunk 6]
```diff
@@ -5,17 +5,17 @@
     habitats" (Final) [Data set].
     Zenodo. https://doi.org/10.5281/zenodo.1255218
 """
-from enum import Enum, auto
+from __future__ import annotations
```
Comment (MylesBartlett) [inline conduit/data/datasets/audio/ecoacoustics.py:8]:
```suggestion
```
annotations should no longer be used as it results in errors when attributes inherited by attrs classes are of types that needs to be imported

File: conduit/data/datamodules/audio/base.py (hunk 1 above)
Comment (DavidHurst) [reply conduit/data/datamodules/audio/base.py:24]: Done✔️

File: conduit/data/datasets/audio/ecoacoustics.py (hunk 6 above)
Comment (DavidHurst) [reply conduit/data/datasets/audio/ecoacoustics.py:8]: Got it, weird, some of these changes I already have locally/on my fork.

**Review by DavidHurst — APPROVED** [review]
All remedied/incorporated.

File: conduit/data/datasets/audio/ecoacoustics.py (hunk 5 above)
Comment (thomkeh) [reply conduit/data/datasets/audio/ecoacoustics.py:97]: I was thinking that `str(preprocessing_transform)` will show more information than just the class name -- like FFT windows size -- (and so allows for more disambiguation), but it doesn't actually seem so, so the approach you're suggesting here should be fine.

File: conduit/types.py [hunk 7]
```diff
@@ -1,4 +1,5 @@
-from enum import Enum
+from __future__ import annotations
```
Comment (thomkeh) [inline conduit/types.py:1]:
```suggestion
from __future__ import annotations
```

File: conduit/data/datamodules/audio/base.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+@attr.define(kw_only=True)
+class CdtAudioDataModule(CdtDataModule):
+
+    root: Union[str, Path] = attr.field(kw_only=False)
+
+    @property
+    @abstractmethod
+    def _base_augmentations(self) -> A.Compose:
+        ...
```
Comment (thomkeh) [inline conduit/data/datamodules/audio/base.py:22]:
```suggestion
```

File: conduit/data/datamodules/audio/ecoacoustics.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+    @property  # type: ignore[misc]
+    @implements(CdtAudioDataModule)
+    def _base_augmentations(self) -> A.Compose:
+        return A.Compose(
+            [
+                A.Resize(self.image_size, self.image_size),
+            ]
+        )
+
```
Comment (thomkeh) [inline conduit/data/datamodules/audio/ecoacoustics.py:42]:
```suggestion
```

**Comment by thomkeh** [comment]
Please review, Myles, and merge if it's all fine.

File: conduit/data/datasets/audio/ecoacoustics.py [hunk 10]
```diff
@@ -5,17 +5,16 @@
     habitats" (Final) [Data set].
     Zenodo. https://doi.org/10.5281/zenodo.1255218
 """
-from enum import Enum, auto
 import math
 from os import mkdir
 from pathlib import Path
+import re
```
Comment (olliethomas) [inline conduit/data/datasets/audio/ecoacoustics.py:11]:
```suggestion
```
Unused

**Review by MylesBartlett — APPROVED** [review]

**Pull request merged**